arches/app/src/components/ControlledListManager/EditLabel.vue

```diff
@@ -0,0 +1,167 @@
+<script setup lang="ts">
+import arches from "arches";
+import { computed, ref } from "vue";
+import { useGettext } from "vue3-gettext";
+
+import Button from "primevue/button";
+import Dialog from "primevue/dialog";
+import Dropdown from "primevue/dropdown";
+import InputText from "primevue/inputtext";
+import { useToast } from "primevue/usetoast";
+
+import { upsertLabel } from "@/components/ControlledListManager/api.ts";
+
+import type {
+    ControlledListItem,
+    Label,
+    LanguageMap,
+} from "@/types/ControlledListManager.d";
+
+const props: {
+    item: ControlledListItem;
+    header: string;
+    languageMap: LanguageMap;
+    label: Label;
+    // updates don't need an insert callback
+    onInsert: null | ((label: Label) => Promise<Label>);
+} = defineProps(["item", "header", "languageMap", "label", "onInsert"]);
+
+const value = ref(props.label.value);
+const language = ref(props.label.language);
+
+const visible = defineModel<boolean>({ required: true });
+
+const languageOptions = computed(() => {
+    if (!props.languageMap) {
+        return [];
+    }
+    return (
+        Object.entries(props.languageMap).map(([code, label]) => {
+            return {
+                label,
+                code,
+            };
+        })
+    );
+});
+
+const toast = useToast();
+const { $gettext } = useGettext();
+const staticItemLabel = $gettext("Item Label");
+const staticLanguageLabel = $gettext("Language");
```

Comment:
why not inline?

Reply:
I found that `$gettext()` executes far more often in the templates than in the script tag, understandably. It's one of the things I'm concerned about most right now. If you checkout this branch, go to the controlled list editor, open the dev console, and drag the splitter handle around, you'll find several thousand of warnings logged for missing language keys. The UI also becomes sluggish, or at least it did before I started moving things into the script tag. And this is on an M2 chip macintosh!

I'm not saying _every_ `gettext()` call needs to be in the setup, but I'd suggest leaving it up to the implementer, unless there's a forseeable peformance impact, e.g. from the splitter panel primevue component as I just mentioned, in which case I'd probably insist on putting it up in the setup.

🎥 :

https://github.com/archesproject/arches/assets/38668450/6e6f40bb-6000-45ce-add0-44b95c759b64

Reply:
holy re-renderings batman! Nice 🎣